packages/appium/test/e2e/driver.e2e.spec.js

```diff
@@ -536,6 +536,68 @@ describe('FakeDriver via HTTP', function () {
       }
     });
   });
+
+  describe('Bidi protocol', function () {
+    withServer();
+    const capabilities = {...caps, webSocketUrl: true};
+
+    it('should respond with bidi specific capability when a driver supports it', async function () {
+      let driver = await wdio({...wdOpts, capabilities});
+      try {
+        should.exist(driver.capabilities.webSocketUrl);
+      } finally {
+        await driver.deleteSession();
+      }
+    });
+
+    it('should interpret the bidi protocol and let the driver handle it by command', async function () {
+      let driver = await wdio({...wdOpts, capabilities});
+      try {
+        should.not.exist(await driver.getUrl());
+
+        await driver.browsingContextNavigate({
+          context: 'foo',
+          url: 'https://appium.io',
+          wait: 'complete',
+        });
+        await driver.getUrl().should.eventually.eql('https://appium.io');
+      } finally {
+        await driver.deleteSession();
+      }
+    });
+
+    it('should be able to subscribe and unsubscribe to bidi events', async function () {
+      let driver = await wdio({...wdOpts, capabilities});
+      let collectedEvents = [];
+      try {
+        // asyncrhonously start our listener
+        driver.on('message', (data) => {
+          const {method, params} = JSON.parse(data.toString('utf8'));
+          if (method === 'clock.currentTime') {
+            collectedEvents.push(params.time);
+          }
+        });
+
+        // wait for some time to be sure clock events have happened, and assert we don't receive
+        // any yet
+        await B.delay(750);
```

Comment:
can we put a timer instead of the hardcoded delay?

Reply:
I'm not sure it would be worth the extra complexity in a test. The important thing is to wait for longer than the clock interval (500ms).